liquid-prelude/src/Language/Haskell/Liquid/Bag.hs

```diff
@@ -44,7 +44,7 @@ put k m = M.insert k (1 + get k m) m
 union :: (Ord k) => Bag k -> Bag k -> Bag k
 union m1 m2 = M.union m1 m2
 
-{-@ thm_emp :: x:k -> xs:Bag k ->  { Language.Haskell.Liquid.Bag.empty /= put x xs }  @-}
+{-@ thm_emp :: x:k -> xs:Bag k ->  { Map_select xs x >= 0 => Language.Haskell.Liquid.Bag.empty /= put x xs }  @-}
```

Comment:
@ranjitjhala, I believe this extra hypothesis is needed, right? Otherwise, how could the SMT solver see that `put x xs` doesn't turn an entry from `-1` to `0`.

I would have expected this to fail before the port, though.

Reply:
I do not see that... Put always returns a non empty bag and this is what the theorem says. 
The domain of put's entry could be express in a different theorem.

Reply:
> Put always returns a non empty bag and this is what the theorem says.

In the representation, an empty Map has all entries set to 0.

Does `put 1 m` produce a map with some entry greater than 0? It depends on what `m` is.

What if I give you the map `Map_store (Map_default 0) 1 (-1)`? It looks like in this case `put 1 m` would produce a map where all entries are 0.

We know, this implementation has an invariant where all maps representing a bag only stores non-negative values. But this invariant is not expressed in this specification of bags.

Does that explain it?